hbase-server/src/main/java/org/apache/hadoop/hbase/master/MetaRegionLocationCache.java

```diff
@@ -0,0 +1,237 @@
+    int replicaId = watcher.getZNodePaths().getMetaReplicaIdFromPath(path);
+    RetryCounter retryCounter = retryCounterFactory.create();
+    HRegionLocation location = null;
+    while (retryCounter.shouldRetry()) {
+      try {
+        if (opType == ZNodeOpType.DELETED) {
+          if (!ZKUtil.watchAndCheckExists(watcher, path)) {
+            // The path does not exist, we've set the watcher and we can break for now.
+            break;
+          }
+          // If it is a transient error and the node appears right away, we fetch the
+          // latest meta state.
+        }
+        location = getMetaRegionLocation(replicaId);
+        break;
+      } catch (KeeperException e) {
+        LOG.debug("Error getting meta location for path {}", path, e);
+        if (!retryCounter.shouldRetry()) {
+          LOG.warn("Error getting meta location for path {}. Retries exhausted.", path, e);
+          break;
+        }
+        try {
+          retryCounter.sleepUntilNextRetry();
+        } catch (InterruptedException ie) {
+          Thread.currentThread().interrupt();
+          return;
+        }
+      }
+    }
+    if (location == null) {
+      cachedMetaLocations.remove(replicaId);
+      return;
+    }
+    cachedMetaLocations.put(replicaId, location);
+  }
+
+  /**
+   * @return Optional list of HRegionLocations for meta replica(s), null if the cache is empty.
+   *
+   */
+  public Optional<List<HRegionLocation>> getMetaRegionLocations() {
+    ConcurrentNavigableMap<Integer, HRegionLocation> snapshot =
+        cachedMetaLocations.tailMap(cachedMetaLocations.firstKey());
+    if (snapshot.isEmpty()) {
+      // This could be possible if the master has not successfully initialized yet or meta region
+      // is stuck in some weird state.
+      return Optional.empty();
+    }
+    List<HRegionLocation> result = new ArrayList<>();
+    // Explicitly iterate instead of new ArrayList<>(snapshot.values()) because the underlying
+    // ArrayValueCollection does not implement toArray().
+    snapshot.values().forEach(location -> result.add(location));
+    return Optional.of(result);
+  }
+
+  /**
+   * Helper to check if the given 'path' corresponds to a meta znode. This listener is only
+   * interested in changes to meta znodes.
+   */
+  private boolean isValidMetaZNode(String path) {
+    return watcher.getZNodePaths().isAnyMetaReplicaZNode(path);
+  }
+
+  @Override
+  public void nodeCreated(String path) {
+    updateMetaLocation(path, ZNodeOpType.CREATED);
+  }
+
+  @Override
+  public void nodeDeleted(String path) {
+    updateMetaLocation(path, ZNodeOpType.DELETED);
+  }
+
+  @Override
+  public void nodeDataChanged(String path) {
+    updateMetaLocation(path, ZNodeOpType.CHANGED);
+  }
+
+  @Override
+  public void nodeChildrenChanged(String path) {
+    if (!path.equals(watcher.getZNodePaths().baseZNode)) {
+      return;
+    }
+    // Can get triggered for *any* children change, but that is OK. It does not happen once the
```

Comment:
So any znode under `/hbase` shows up or disappears and we re-build the cache? Seems like this would cause a lot of unnecessary ZK chatter. I really wish the meta replica list was under a node of its own.

Reply:
> I really wish the meta replica list was under a node of its own.

I was thinking the same.

Taking a step back, this whole logic is implemented to handle the following scenario. Consider a timeline of events (t n+1 > t n)

t1: standby master starts (populateInitialMetaLocations() runs in c'tor). No meta znodes, so nothing is populated in the cache
t2: Active master starts
t3: meta znodes created.

At this point standby's meta cache is stale and it hasn't registered any watchers, so nothing is notified and will remain a zombie forever.

So to fix this problem, populateInitialMetaLocations() registers a watcher on the *base* znode *and* also gets the meta znodes. So if the meta znodes turn out to be empty, we at least have a watcher on the base znode when something gets populated later. So the new timeline would be.

t1: standby master starts. No meta znodes, *but* registered a watcher on /hbase
t2: active master starts
t3: meta znodes created
t4: standby gets a notification that /hbase has changed and it populates the cache.

Do remember that we *do not* renew the watcher. So t4 is the last nodeChildrenChanged() event.

Now the only problem is in the following sequence of actions.

t1: active master starts
t2. meta znodes created.
t3: standby master starts, reads the meta znodes *but* also has a watcher on /hbase.

Now in the future if some children of /hbase changes, we have an extraneous invalidate of cache, but thats ok since that is the last time it is going to happen.

Now one may ask, why not do it the following way.

1. meta_znodes = getMetaZnodes();
2. if (meta znodes == empty) {
    register a watcher on /hbase
3. }

Unfortunately this is not atomic. Between (1) and (2), the meta znodes could be created and standby could remain stale for a while until some children of /hbase changes.

Sorry for the long post. Hope this explains my line of thought. Feel free to point anything that I'm missing here. :-)

Reply:
One additional thought, what if some other thread registers a watcher on the base znode and the cache thread processes them (guess this is possible). should we make populateInitialMetaLocations() idempotent? That solves both the problems I guess.

Reply:
Exactly, these are the kinds of timelines I've been reasoning through.

All of these operations seem prone to data races because we have to reach back out to zookeeper to retrieve updated information. This is why I was designing a system that made use of the znode version information so that state updates at least always march forward, but that's apparently not available to us.

For what it's worth, what your building is essentially a [Node Cache](https://curator.apache.org/curator-recipes/node-cache.html) on each `meta-region-server` node, or maybe a [Path Cache](https://curator.apache.org/curator-recipes/path-cache.html) on a subset of the `/hbase` node. I'm not suggesting we introduce a dependency on Curator, but maybe that project has some useful ideas/examples/test facilities that would help here.

Reply:
Oh thanks for the pointers, didn't know about NodeCache and PatchCache. They'd definitely simplify the entire ZK state management in HBase.